README.rst
=============================
CAE app
=============================

.. image:: https://badge.fury.io/py/django-cae.svg
    :target: https://badge.fury.io/py/django-cae

.. image:: https://travis-ci.org/adevolutio/django-cae.svg?branch=master
    :target: https://travis-ci.org/adevolutio/django-cae

.. image:: https://codecov.io/gh/adevolutio/django-cae/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/adevolutio/django-cae

Classificação Portuguesa das Actividades Económicas

Documentation
-------------

The full documentation is at https://django-cae.readthedocs.io.

Quickstart
----------

Install CAE app::

    pip install django-cae

Add it to your `INSTALLED_APPS`:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'cae.apps.CaeConfig',
        ...
    )

Add CAE app's URL patterns:

.. code-block:: python

    from cae import urls as cae_urls


    urlpatterns = [
        ...
        url(r'^', include(cae_urls)),
        ...
    ]

Features
--------

* TODO

Running Tests
-------------

Does the code actually work?

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install tox
    (myenv) $ tox


Development commands
---------------------

::

    pip install -r requirements_dev.txt
    invoke -l


Credits
-------

Tools used in rendering this package:

*  Cookiecutter_
*  `cookiecutter-djangopackage`_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`cookiecutter-djangopackage`: https://github.com/pydanny/cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2021-05-04)
++++++++++++++++++

* First release on PyPI.


tests/settings.py
"""
Django settings for example project.

Generated by Cookiecutter Django Package
"""

import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.9/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "3=di3ek_9n!j*hu)kl6#qzme#cb)+2+(6$d8303@arj28enorz"

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    'cae',
    

    # if your app has other dependencies that need to be added to the site
    # they should be added here
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'tests.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates'), ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'tests.wsgi.application'

# Database
# https://docs.djangoproject.com/en/1.9/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Password validation
# https://docs.djangoproject.com/en/1.9/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# Internationalization
# https://docs.djangoproject.com/en/1.9/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.9/howto/static-files/

STATIC_URL = '/static/'


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_django-cae
------------

Tests for `django-cae` models module.
"""

from django.test import TestCase

from cae import models


class TestCae(TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


tests/example/urls.py
"""example URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/1.9/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  url(r'^$', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  url(r'^$', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.conf.urls import url, include
    2. Add a URL to urlpatterns:  url(r'^blog/', include('blog.urls'))
"""
from django.conf.urls import url, include
from django.contrib import admin


urlpatterns = [
    url(r'^admin/', admin.site.urls),
    url(r'', include('cae.urls', namespace='cae')),
]


tasks.py
import os
import webbrowser

from invoke import task


def open_browser(path):
    try:
        from urllib import pathname2url
    except:
        from urllib.request import pathname2url
    webbrowser.open("file://" + pathname2url(os.path.abspath(path)))


@task
def clean_build(c):
    """
    Remove build artifacts
    """
    c.run("rm -fr build/")
    c.run("rm -fr dist/")
    c.run("rm -fr *.egg-info")


@task
def clean_pyc(c):
    """
    Remove python file artifacts
    """
    c.run("find . -name '*.pyc' -exec rm -f {} +")
    c.run("find . -name '*.pyo' -exec rm -f {} +")
    c.run("find . -name '*~' -exec rm -f {} +")


@task
def coverage(c):
    """
    check code coverage quickly with the default Python
    """
    c.run("coverage run --source django-cae runtests.py tests")
    c.run("coverage report -m")
    c.run("coverage html")
    c.run("open htmlcov/index.html")


@task
def docs(c):
    """
    Build the documentation and open it in the browser
    """
    c.run("rm -f docs/django-cae.rst")
    c.run("rm -f docs/modules.rst")
    c.run("sphinx-apidoc -o docs/ cae")

    c.run("sphinx-build -E -b html docs docs/_build")
    open_browser(path='docs/_build/html/index.html')


@task
def test_all(c):
    """
    Run tests on every python version with tox
    """
    c.run("tox")


@task
def clean(c):
    """
    Remove python file and build artifacts
    """
    clean_build(c)
    clean_pyc(c)


@task
def unittest(c):
    """
    Run unittests
    """
    c.run("python manage.py test")


@task
def lint(c):
    """
    Check style with flake8
    """
    c.run("flake8 django-cae tests")


@task(help={'bumpsize': 'Bump either for a "feature" or "breaking" change'})
def release(c, bumpsize=''):
    """
    Package and upload a release
    """
    clean(c)
    if bumpsize:
        bumpsize = '--' + bumpsize

    c.run("bumpversion {bump} --no-input".format(bump=bumpsize))

    import django_cae
    c.run("python setup.py sdist bdist_wheel")
    c.run("twine upload dist/*")

    c.run('git tag -a {version} -m "New version: {version}"'.format(version=django_cae.__version__))
    c.run("git push --tags")
    c.run("git push origin master")


cae/models.py
# -*- coding: utf-8 -*-
from django.db import models
from django.utils.translation import ugettext as _


class Section(models.Model):
    code = models.CharField(_('Código'), max_length=1, primary_key=True)
    name = models.CharField(_('Designação'), max_length=128)

    def __str__(self):
        return "%s: %s" % (self.code, self.name)

    class Meta:
        verbose_name = _('Secção')
        verbose_name_plural = _('Secções')
        db_table = 'cae_section'


class Division(models.Model):
    code = models.SmallIntegerField(_('Código'), primary_key=True)
    name = models.CharField(_('Designação'), max_length=128)
    section = models.ForeignKey(Section, verbose_name=_('Secção'), on_delete=models.CASCADE)

    def __str__(self):
        return "%s: %s" % (self.formatted_code, self.name)

    class Meta:
        verbose_name = _('Divisão')
        verbose_name_plural = _('Divisões')
        db_table = 'cae_division'

    @property
    def formatted_code(self):
        return "%02d" % self.code


class CAE(models.Model):
    level = models.SmallIntegerField(_('Nível'))
    code = models.IntegerField(_('Código'))
    name = models.CharField(_('Designação'), max_length=255)

    def __str__(self):
        return "%s - %s" % (self.formatted_code, self.name)

    class Meta:
        verbose_name = _('CAE')
        verbose_name_plural = _('CAE\'s')
        db_table = 'cae_cae'
        unique_together = ('level', 'code')

    @property
    def formatted_code(self):
        if self.level == 3:
            return "%03d" % self.code
        elif self.level == 4:
            return "%04d" % self.code
        else:
            return "%05d" % self.code


cae/migrations/0001_initial.py
# Generated by Django 3.2.1 on 2021-05-04 15:44

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Section',
            fields=[
                ('code', models.CharField(max_length=1, primary_key=True, serialize=False, verbose_name='Código')),
                ('name', models.CharField(max_length=128, verbose_name='Designação')),
            ],
            options={
                'verbose_name': 'Secção',
                'verbose_name_plural': 'Secções',
                'db_table': 'cae_section',
            },
        ),
        migrations.CreateModel(
            name='Division',
            fields=[
                ('code', models.SmallIntegerField(primary_key=True, serialize=False, verbose_name='Código')),
                ('name', models.CharField(max_length=128, verbose_name='Designação')),
                ('section', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='cae.section', verbose_name='Secção')),
            ],
            options={
                'verbose_name': 'Divisão',
                'verbose_name_plural': 'Divisões',
                'db_table': 'cae_division',
            },
        ),
        migrations.CreateModel(
            name='CAE',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('level', models.SmallIntegerField(verbose_name='Nível')),
                ('code', models.IntegerField(verbose_name='Código')),
                ('name', models.CharField(max_length=255, verbose_name='Designação')),
            ],
            options={
                'verbose_name': 'CAE',
                'verbose_name_plural': "CAE's",
                'db_table': 'cae_cae',
                'unique_together': {('level', 'code')},
            },
        ),
    ]
